department-of-veterans-affairs/abd-vro
# 1676: Implement Kafka Client to Connect to Mock Kafka Service and Forward BIE Payload to RabbitMQ

## What was the problem?
Need a microservice that connects to Kafka mock and subscribes to topics. When messages are received on subscribed topics, those messages are used as a payload in a message to RabbitMQ exchange corresponding to the Kafka topic.

Associated tickets or Slack threads:
- #1676
- Dependent on 
  - #1675 

## How does this fix it?
Creates the client and a way to run and test it.

## How to test this PR
First clean, build, and run the platform docker.
```
source scripts/setenv.sh
# Build all containers
./gradlew clean 
./gradlew build 
./gradlew docker
# Start the relevant containers
./gradlew :dockerComposeUp
```
Note: you may have to restart the RabbitMQ service. This is a known issue on M series Macs.

### To test microservice outside of docker:
1. Build and run the Kafka mock. The first line builds the Kafka mock, and the second line runs the mock exposing the docker port 9092.
```
./gradlew -p mocks :mock-bie-kafka:docker 
docker run --rm -it -p 9092:9092 va/vro_mocks-mock-bie-kafka
```


2. Open a separate terminal and run the kcat command to create a topic:
```kcat -b 127.0.0.1:9092 -P -t TST_CONTENTION_BIE_CONTENTION_ASSOCIATED_TO_CLAIM_V02```

3. Run the bootrun task for the microservice
```./gradlew :svc-bie-kafka:bootRun```

4. Send a message on the topic using kcat:
```kcat -b 127.0.0.1:9092 -P -t TST_CONTENTION_BIE_CONTENTION_ASSOCIATED_TO_CLAIM_V02```
Type your message and press CTRL-D to end the message. Each line will send a message.

### Testing with docker:
1. Build and run the kafka mock
```
./gradlew -p mocks :mock-bie-kafka:docker 
COMPOSE_PROFILES="kafka" ./gradlew -p mocks :dockerComposeUp
```
2. Open terminal within the mock container and create the topic:
```
kafka-topics.sh --bootstrap-server localhost:9092 --topic TST_CONTENTION_BIE_CONTENTION_ASSOCIATED_TO_CLAIM_V02 --create --partitions 3 --replication-factor 1
```
3. Run the microservice application:
```COMPOSE_PROFILES="kafka" ./gradlew :app:dockerComposeUp```
4. In the mock Kafka container, open terminal and send a message on the topic:
```
kafka-console-producer.sh --bootstrap-server localhost:9092 --topic TST_CONTENTION_BIE_CONTENTION_ASSOCIATED_TO_CLAIM_V02
``` 
Type your message and press CTRL-C to end the message. Each line will send a message.
For more information on the kafka CLI in bitnami image, see [here](https://www.conduktor.io/kafka/kafka-cli-tutorial/)

### Verification:

1. Verify message was received in svs-bie-kafka log output
2. Verify message was converted to BieMessagePayload and sent and received to RabbitMQ by opening the queue [bie-events-contention-associated](http://localhost:15672/#/queues/%2F/bie-events-contention-associated). 
3. Regardless of using docker for the test, you should see something similar to the picture below:
<img width="1000" alt="Screen Shot 2023-06-29 at 8 30 16 AM" src="https://github.com/department-of-veterans-affairs/abd-vro/assets/135860892/30464747-6556-45a6-8ea5-3441f4e4b782">

## Notes about work:
This ticket is part of epic #1642
